### ISSUE TITLE
xPackage uninstall with EXE installer fails

### ISSUE THREAD

**Issue by Bertk** [dsccommunity/xPSDesiredStateConfiguration#200]
xPackage starts msiexec.exe instead of the EXE file with uninstall parameters and therefore replacing an old version with a newer package is not possible.

```
VERBOSE: [DESKTOP-RJC225Q]: [[xPackage]XXXsetup] The path extension was .exe
VERBOSE: [DESKTOP-RJC225Q]: [[xPackage]XXXsetup] Package configuration starting
VERBOSE: [DESKTOP-RJC225Q]: [[xPackage]XXXsetup] The binary is an EXE
VERBOSE: [DESKTOP-RJC225Q]: [[xPackage]XXXsetup] Starting C:\Windows\system32\msiexec.exe
with /x NameOfPackage /quiet /norestart /S
VERBOSE: [DESKTOP-RJC225Q]: [[xPackage]XXXsetup] Starting process
C:\Windows\system32\msiexec.exe with arguments /x NameOfPackage /quiet /norestart /S
```

The installations is OK.

```
VERBOSE: [DESKTOP-RJC225Q]: [[xPackage]XXXsetup] The binary is an EXE
VERBOSE: [DESKTOP-RJC225Q]: [[xPackage]XXXsetup] Starting C:\product_addons\XXX-setup.exe
with /S
VERBOSE: [DESKTOP-RJC225Q]: [[xPackage]XXXsetup] Starting process
C:\product_addons\XXX-setup.exe with arguments /S
VERBOSE: [DESKTOP-RJC225Q]: [[xPackage]XXXsetup] Perform operation 'Invoke CimMethod' with
following parameters, ''methodName' = GetServerFeature,'className' = MSFT_ServerManagerTasks,'namespaceName' =
root\microsoft\windows\servermanager'.
```

**Comment by pshamus** [comment]
I just ran into this issue but with the built-in MSFT_PackageResource. I was thinking a solution could be to have two new parameters: UninstallPath and UninstallArguments. That would take precedence when Ensure = 'Absent'. Thoughts?

**Comment by HansOMartinsen** [comment]
Hi @mbreakey3 . This issue is been assigned since Feb 16. Are you going to change some of the logic here? The reason I ask is that I have met with the same problem, and have discovered some strangeness in the code I am curious about. In MSFT_xPackageResource.psm1, line 317 (version 44a89e8 on 8 Apr):

if (-not ($fileExtension -eq '.msi' -and $Ensure -eq 'Absent'))

Question 1:
I  cannot wrap my head around why the logic is for that If statement. As far as I can see there are four states that are possible:
1: The file is an MSI and the required Ensure is Present = True
2: The file is an MSI and the required Ensure is Absent = False
3: The file is NOT an MSI and the required Ensure is Present = True
4: The file is NOT an MSI and the required Ensure is Absent = True
Is the purpose to say that all exe files must be downloaded if they are being removed or installed and MSI if it is being installed?

In that case there is a flaw in the way the If and ElseIf statement in line 319 and 336 is handling the download. It will only cover state 1 and 3 (not 4  as expected), when the file is downloaded from a Web path. I would suggest to remove "-and $Ensure -eq 'Present' " from line 336 as to not and up with an error on line 429. Should I create a pull request for this, or are you looking at other ways to solve it?

**Comment by mbreakey3** [comment]
Hi @HansOMartinsen, Yes - the current xPacakge resource is broken for EXE. There's certainly some strange logic in parts of it that don't make sense to me either. I originally was planning on splitting this resource into two resources: MsiPackage and ExePackage since the two types are dealt with very differently and really should be in separate resources. Unfortunately, I was only able to complete MsiPackage (which is now a High Quality resource in [PSDscResources](https://github.com/PowerShell/PSDscResources/tree/dev/DscResources/MSFT_MsiPackage)) before I was moved over to a different task. I will un-assign myself from this issue, as it is misleading. 

Long story short - we know that this resource is broken, we just don't have the bandwidth internally to be able to fix it right now. There's quite a bit of logic that needs to be added/changed to ensure this resource works for EXE packages, but you or anyone else is welcome to open a PR with fixes to the problem.

**Comment by HansOMartinsen** [comment]
Thanks for the clear and candid response. I will look into if there are time to look at this from my end.
